src/main/kotlin/translations/identification/TranslationExternalAnnotationsArtifactsResolver.kt

```diff
@@ -0,0 +1,130 @@
+ * by the Free Software Foundation, version 3.0 only.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.demonwav.mcdev.translations.identification
+
+import com.demonwav.mcdev.util.invokeAndWait
+import com.demonwav.mcdev.util.invokeLater
+import com.intellij.codeInsight.ExternalAnnotationsArtifactsResolver
+import com.intellij.codeInsight.externalAnnotation.location.AnnotationsLocation
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.PathManager
+import com.intellij.openapi.application.runWriteAction
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.roots.AnnotationOrderRootType
+import com.intellij.openapi.roots.impl.libraries.LibraryEx
+import com.intellij.openapi.roots.libraries.Library
+import com.intellij.openapi.roots.libraries.ui.OrderRoot
+import com.intellij.openapi.roots.ui.configuration.libraryEditor.ExistingLibraryEditor
+import com.intellij.openapi.util.io.FileUtil
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.openapi.vfs.VirtualFileManager
+import org.jetbrains.concurrency.AsyncPromise
+import org.jetbrains.concurrency.Promise
+import org.jetbrains.concurrency.resolvedPromise
+
+class TranslationExternalAnnotationsArtifactsResolver : ExternalAnnotationsArtifactsResolver {
+    override fun resolve(project: Project, library: Library, mavenId: String?): Boolean {
+        if (!Util.isOurFakeMavenLocation(mavenId)) {
+            return false
+        }
+
+        return invokeAndWait {
+            doResolve(library)
+        }
+    }
+
+    override fun resolve(project: Project, library: Library, annotationsLocation: AnnotationsLocation): Boolean {
+        if (annotationsLocation != Util.fakeMavenLocation) {
+            return false
+        }
+        return invokeAndWait {
+            doResolve(library)
+        }
+    }
+
+    override fun resolveAsync(project: Project, library: Library, mavenId: String?): Promise<Library> {
+        if (!Util.isOurFakeMavenLocation(mavenId)) {
+            return resolvedPromise(library)
+        }
+
+        val promise = AsyncPromise<Library>()
+        ApplicationManager.getApplication().executeOnPooledThread {
+            invokeLater {
+                doResolve(library)
+                promise.setResult(library)
+            }
+        }
+
+        return promise
+    }
+
+    private fun doResolve(library: Library): Boolean {
+        if (library !is LibraryEx || library.isDisposed) {
+            return false
+        }
+
+        return runWriteAction {
```

Comment:
Make sure to check that the library is not disposed inside the write action.

Reply:
I copied from `ExternalAnnotationsRepositoryResolver` again here, I don't think it's possible for the library to be disposed between these two points because `runWriteAction` is sync and we're on the EDT

Reply:
Right, sounds good then 👍